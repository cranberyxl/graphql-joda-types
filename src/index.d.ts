export interface Type<T> extends Function {
  new (...args: any[]): T;
}
export declare const LocalDate: any;
export declare const LocalDateTime: any;
export declare const LocalTime: any;
export declare const ZonedDateTime: any;
export declare const ZoneId: any;
