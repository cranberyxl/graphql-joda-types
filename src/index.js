// @flow
import { GraphQLScalarType, Kind } from 'graphql';
import type { GraphQLScalarTypeConfig } from 'graphql';
import * as joda from 'js-joda';

const jodaToGraphql = <T>(
  name: string,
  description: string,
  jodaType: Class<T>, // eslint-disable-line no-undef
  parser: string => T
) => {
  const parse = (value: string): T => {
    try {
      return parser(value);
    } catch (e) {
      throw new TypeError(
        `${name} can not represent an invalid string ${value}`
      );
    }
  };

  const config: GraphQLScalarTypeConfig<T, string> = {
    name,
    description,
    serialize: (value): string => {
      if (value instanceof jodaType) {
        return value.toString();
      }
      if (typeof value === 'string') {
        parse(value); // will throw if it's invalid
        return value;
      }
      throw new TypeError(
        `${name} can not represent invalid type ${JSON.stringify(value)}`
      );
    },
    parseValue: (value): T => {
      if (typeof value === 'string') {
        return parse(value);
      }
      throw new TypeError(
        `${name} can not represent non-string type ${JSON.stringify(value)}`
      );
    },
    parseLiteral: (ast): T => {
      if (ast.kind !== Kind.STRING) {
        throw new TypeError(
          `${name} can not represent non-string type ${String(
            ast.value != null ? ast.value : null
          )}`
        );
      }
      return parse(ast.value);
    },
  };

  return new GraphQLScalarType(config);
};

export const LocalDate = jodaToGraphql(
  'LocalDate',
  'A date string, such as 2018-07-01, serialized in ISO8601 format',
  joda.LocalDate,
  joda.LocalDate.parse
);
export const LocalDateTime = jodaToGraphql(
  'LocalDateTime',
  'A date and time, such as 2018-07-01T12:00:00, without a timezone, serialized in ISO8601 format',
  joda.LocalDateTime,
  joda.LocalDateTime.parse
);
export const LocalTime = jodaToGraphql(
  'LocalTime',
  'A time string, such as 12:00:00, serialized in ISO8601 format',
  joda.LocalTime,
  joda.LocalTime.parse
);
export const ZonedDateTime = jodaToGraphql(
  'ZonedDateTime',
  'A specific moment in time, such as 2018-07-01T12:00:00-04:00, serialized in ISO8601 format',
  joda.ZonedDateTime,
  joda.ZonedDateTime.parse
);
export const ZoneId = jodaToGraphql(
  'ZoneId',
  'A timezone ID from the IANA timezone database, such as America/New_York or UTC',
  joda.ZoneId,
  joda.ZoneId.of
);
