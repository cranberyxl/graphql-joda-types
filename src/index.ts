import { GraphQLScalarType, Kind } from 'graphql';
import * as joda from '@js-joda/core';

export const LocalDate = new GraphQLScalarType({
  name: 'LocalDate',
  description:
    'A date string, such as 2018-07-01, serialized in ISO8601 format',
  serialize: (value): string => {
    if (value instanceof joda.LocalDate) {
      return value.toString();
    }
    if (typeof value === 'string') {
      joda.LocalDate.parse(value); // will throw if it's invalid
      return value;
    }
    throw new TypeError(
      `${name} can not represent invalid type ${JSON.stringify(value)}`
    );
  },
  parseValue: (value): joda.LocalDate => {
    if (typeof value === 'string') {
      return joda.LocalDate.parse(value);
    }
    throw new TypeError(
      `${name} can not represent non-string type ${JSON.stringify(value)}`
    );
  },
  parseLiteral: (ast): joda.LocalDate => {
    if (ast.kind !== Kind.STRING) {
      throw new TypeError(
        `${name} can not represent non-string type ${JSON.stringify(ast)}`
      );
    }
    return joda.LocalDate.parse(ast.value);
  },
});

export const LocalDateTime = new GraphQLScalarType({
  name: 'LocalDateTime',
  description:
    'A date and time, such as 2018-07-01T12:00:00, without a timezone, serialized in ISO8601 format',
  serialize: (value): string => {
    if (value instanceof joda.LocalDateTime) {
      return value.toString();
    }
    if (typeof value === 'string') {
      joda.LocalDateTime.parse(value); // will throw if it's invalid
      return value;
    }
    throw new TypeError(
      `${name} can not represent invalid type ${JSON.stringify(value)}`
    );
  },
  parseValue: (value): joda.LocalDateTime => {
    if (typeof value === 'string') {
      return joda.LocalDateTime.parse(value);
    }
    throw new TypeError(
      `${name} can not represent non-string type ${JSON.stringify(value)}`
    );
  },
  parseLiteral: (ast): joda.LocalDateTime => {
    if (ast.kind !== Kind.STRING) {
      throw new TypeError(
        `${name} can not represent non-string type ${JSON.stringify(ast)}`
      );
    }
    return joda.LocalDateTime.parse(ast.value);
  },
});

export const LocalTime = new GraphQLScalarType({
  name: 'LocalTime',
  description: 'A time string, such as 12:00:00, serialized in ISO8601 format',
  serialize: (value): string => {
    if (value instanceof joda.LocalTime) {
      return value.toString();
    }
    if (typeof value === 'string') {
      joda.LocalTime.parse(value); // will throw if it's invalid
      return value;
    }
    throw new TypeError(
      `${name} can not represent invalid type ${JSON.stringify(value)}`
    );
  },
  parseValue: (value): joda.LocalTime => {
    if (typeof value === 'string') {
      return joda.LocalTime.parse(value);
    }
    throw new TypeError(
      `${name} can not represent non-string type ${JSON.stringify(value)}`
    );
  },
  parseLiteral: (ast): joda.LocalTime => {
    if (ast.kind !== Kind.STRING) {
      throw new TypeError(
        `${name} can not represent non-string type ${JSON.stringify(ast)}`
      );
    }
    return joda.LocalTime.parse(ast.value);
  },
});

export const ZonedDateTime = new GraphQLScalarType({
  name: 'ZonedDateTime',
  description:
    'A specific moment in time, such as 2018-07-01T12:00:00-04:00, serialized in ISO8601 format',
  serialize: (value): string => {
    if (value instanceof joda.ZonedDateTime) {
      return value.toString();
    }
    if (typeof value === 'string') {
      joda.ZonedDateTime.parse(value); // will throw if it's invalid
      return value;
    }
    throw new TypeError(
      `${name} can not represent invalid type ${JSON.stringify(value)}`
    );
  },
  parseValue: (value): joda.ZonedDateTime => {
    if (typeof value === 'string') {
      return joda.ZonedDateTime.parse(value);
    }
    throw new TypeError(
      `${name} can not represent non-string type ${JSON.stringify(value)}`
    );
  },
  parseLiteral: (ast): joda.ZonedDateTime => {
    if (ast.kind !== Kind.STRING) {
      throw new TypeError(
        `${name} can not represent non-string type ${JSON.stringify(ast)}`
      );
    }
    return joda.ZonedDateTime.parse(ast.value);
  },
});

export const ZoneId = new GraphQLScalarType({
  name: 'ZoneId',
  description:
    'A timezone ID from the IANA timezone database, such as America/New_York or UTC',
  serialize: (value): string => {
    if (value instanceof joda.ZoneId) {
      return value.toString();
    }
    if (typeof value === 'string') {
      joda.ZoneId.of(value); // will throw if it's invalid
      return value;
    }
    throw new TypeError(
      `${name} can not represent invalid type ${JSON.stringify(value)}`
    );
  },
  parseValue: (value): joda.ZoneId => {
    if (typeof value === 'string') {
      return joda.ZoneId.of(value);
    }
    throw new TypeError(
      `${name} can not represent non-string type ${JSON.stringify(value)}`
    );
  },
  parseLiteral: (ast): joda.ZoneId => {
    if (ast.kind !== Kind.STRING) {
      throw new TypeError(
        `${name} can not represent non-string type ${JSON.stringify(ast)}`
      );
    }
    return joda.ZoneId.of(ast.value);
  },
});

export const Duration = new GraphQLScalarType({
  name: 'Duration',
  description: 'A time-based amount of time, such as "34.5 seconds"',
  serialize: (value): string => {
    if (value instanceof joda.Duration) {
      return value.toString();
    }
    if (typeof value === 'string') {
      joda.Duration.parse(value); // will throw if it's invalid
      return value;
    }
    throw new TypeError(
      `${name} can not represent invalid type ${JSON.stringify(value)}`
    );
  },
  parseValue: (value): joda.Duration => {
    if (typeof value === 'string') {
      return joda.Duration.parse(value);
    }
    throw new TypeError(
      `${name} can not represent non-string type ${JSON.stringify(value)}`
    );
  },
  parseLiteral: (ast): joda.Duration => {
    if (ast.kind !== Kind.STRING) {
      throw new TypeError(
        `${name} can not represent non-string type ${JSON.stringify(ast)}`
      );
    }
    return joda.Duration.parse(ast.value);
  },
});
