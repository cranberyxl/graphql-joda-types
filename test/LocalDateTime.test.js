import { Kind } from 'graphql';
import * as joda from 'js-joda';
import { LocalDateTime } from '../src';

describe('LocalDateTime', () => {
  it('has a description', () => {
    expect(LocalDateTime.description).toMatchSnapshot();
  });

  it('parses valid datetimes', () => {
    expect(LocalDateTime.parseValue('2018-07-01T20:00:00')).toEqual(
      joda.LocalDateTime.of(2018, 7, 1, 20)
    );

    expect(
      LocalDateTime.parseLiteral({
        kind: Kind.STRING,
        value: '2018-07-01T20:00:00',
      })
    ).toEqual(joda.LocalDateTime.of(2018, 7, 1, 20));
  });

  it('rejects invalid datetimes', () => {
    expect(() =>
      LocalDateTime.parseValue('2018-07-01')
    ).toThrowErrorMatchingSnapshot();
  });

  it('parses undefined to undefined', () => {
    expect(LocalDateTime.parseValue(undefined)).toBeUndefined();
  });

  it('serializes dates', () => {
    expect(
      LocalDateTime.serialize(joda.LocalDateTime.of(2018, 7, 1, 20))
    ).toEqual('2018-07-01T20:00');
  });

  it('serializes strings (by passing them through)', () => {
    expect(LocalDateTime.serialize('2018-07-01T20:00:00')).toEqual(
      '2018-07-01T20:00:00'
    );
  });

  it('refuses to serialize strings that are not valid dates', () => {
    expect(() =>
      LocalDateTime.serialize('some garbage')
    ).toThrowErrorMatchingSnapshot();
  });
});
