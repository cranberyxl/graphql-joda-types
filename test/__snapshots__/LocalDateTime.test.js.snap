// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocalDateTime has a description 1`] = `"A date and time, such as 2018-07-01T12:00:00, without a timezone, serialized in ISO8601 format"`;

exports[`LocalDateTime refuses to serialize strings that are not valid dates 1`] = `"LocalDateTime can not represent an invalid string some garbage"`;

exports[`LocalDateTime rejects invalid datetimes 1`] = `"LocalDateTime can not represent an invalid string 2018-07-01"`;
